### PR TITLE
planner: fix the issue that UnionAll didn't handle the range bump case

### LINKED ISSUE #52472: Unexpected SQL exception returned
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
I run the following statement through JDBC with version 8.3.0

```sql
CREATE TABLE t1 ( c1 int, c2 int);
CREATE TABLE t2 ( c1 int unsigned);
INSERT INTO t1 (c1,c2) VALUES (8,4);
INSERT INTO t2 (c1) VALUES (2454396638);
SELECT ca4 AS ca1 FROM (SELECT c2 AS ca4 FROM t1) AS ta2 CROSS JOIN (SELECT c1 FROM t2) AS ta3 UNION ALL SELECT c1 AS ca13 FROM t2; 
```
### 2. What did you expect to see? (Required)
The JDBC should return {2454396638, 4}
### 3. What did you see instead (Required)
A SQLException is returned.
```java
java.sql.SQLException: Out of range value for column 'ca1' : value 2454396638 is not in class java.lang.Integer range
	at org.mariadb.jdbc.internal.com.read.resultset.rowprotocol.RowProtocol.rangeCheck(RowProtocol.java:283)
	at org.mariadb.jdbc.internal.com.read.resultset.rowprotocol.TextRowProtocol.getInternalInt(TextRowProtocol.java:255)
	at org.mariadb.jdbc.internal.com.read.resultset.rowprotocol.TextRowProtocol.getInternalObject(TextRowProtocol.java:900)
	at org.mariadb.jdbc.internal.com.read.resultset.SelectResultSet.getObject(SelectResultSet.java:1159)
	in JdbcHelperImpl.getObject(JdbcHelperImpl.java:353)
```
### 4. What is your TiDB version? (Required)

Release Version: v8.0.0
Edition: Community
Git Commit Hash: 8ba1fa452b1ccdbfb85879ea94b9254aabba2916
Git Branch: HEAD
UTC Build Time: 2024-03-28 14:22:15
GoVersion: go1.21.4
Race Enabled: false
Check Table Before Drop: false
Store: tikv

### PR DESCRIPTION
**Pull request by YangKeao**
### What problem does this PR solve?

Issue Number: close #52472

Problem Summary:

The `UnionAll` should use `types.AggFieldType` to aggregate the fieldTypes. This function will handle the bump of range. For example, the aggregate of signed and unsigned long will give signed long long.

### What changed and how does it work?

Use `types.AggFieldType` instead of `types.MergeFieldType`.

This PR also makes `types.MergeFieldType` private, as it's easy to make such mistakes.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the issue that the content of the result of `union all` may overflow the type.
```

### CODE REVIEW

File: tests/integrationtest/r/explain_easy.result [hunk 1]
```diff
@@ -516,7 +516,7 @@ StreamAgg	1.00	root		funcs:count(1)->Column#22
       │   └─Projection	1.00	root		Column#33, Column#34, Column#35, Column#14, Column#15, Column#16
       │     └─HashAgg	1.00	root		group by:explain_easy.test01.region_id, explain_easy.test01.show_date, explain_easy.test01.stat_date, funcs:firstrow(explain_easy.test01.stat_date)->Column#33, funcs:firstrow(explain_easy.test01.show_date)->Column#34, funcs:firstrow(explain_easy.test01.region_id)->Column#35, funcs:firstrow(explain_easy.test01.stat_date)->Column#14, funcs:firstrow(explain_easy.test01.show_date)->Column#15, funcs:firstrow(explain_easy.test01.region_id)->Column#16, funcs:count(1)->Column#39
       │       └─TableReader	0.01	root		data:Selection
-      │         └─Selection	0.01	cop[tikv]		eq(explain_easy.test01.period, 1), ge(explain_easy.test01.stat_date, 20191202), gt(cast(explain_easy.test01.registration_num, bigint(20) BINARY), 0), le(explain_easy.test01.stat_date, 20191202)
+      │         └─Selection	0.01	cop[tikv]		eq(explain_easy.test01.period, 1), ge(explain_easy.test01.stat_date, 20191202), gt(cast(explain_easy.test01.registration_num, decimal(20,0) BINARY), 0), le(explain_easy.test01.stat_date, 20191202)
```
Comment (YangKeao) [inline tests/integrationtest/r/explain_easy.result:519]: It unions integer literal 0 with the unsigned big int. It's expected to produce new decimal. I've checked that this behavior is compatible with MySQL, and the related tests have been added in `TestIssue52472`

File: pkg/planner/core/casetest/hint/testdata/integration_suite_out.json [hunk 2]
```diff
@@ -786,12 +786,10 @@
...
-          "              ├─Projection 3323.33 mpp[tiflash]  cast(test.t.a, int(11) BINARY)->Column#14, test.t.b->Column#15",
-          "              │ └─Selection 3323.33 mpp[tiflash]  lt(test.t.a, 18), lt(test.t.a, 60)",
-          "              │   └─TableFullScan 10000.00 mpp[tiflash] table:t pushed down filter:empty, keep order:false, stats:pseudo",
-          "              └─Projection 1107.78 mpp[tiflash]  cast(test.t.a, int(11) BINARY)->Column#14, test.t.b->Column#15",
-          "                └─Selection 1107.78 mpp[tiflash]  gt(test.t.b, 1), lt(test.t.a, 60)",
-          "                  └─TableFullScan 10000.00 mpp[tiflash] table:t pushed down filter:empty, keep order:false, stats:pseudo"
+          "              ├─Selection 3323.33 mpp[tiflash]  lt(test.t.a, 18), lt(test.t.a, 60)]",
+          "              │ └─TableFullScan 10000.00 mpp[tiflash] table:t pushed down filter:empty, keep order:false, stats:pseudo]",
+          "              └─Selection 1107.78 mpp[tiflash]  gt(test.t.b, 1), lt(test.t.a, 60)]",
+          "                └─TableFullScan 10000.00 mpp[tiflash] table:t pushed down filter:empty, keep order:false, stats:pseudo]",
```
Comment (YangKeao) [inline pkg/planner/core/casetest/hint/testdata/integration_suite_out.json:792]: It's expected. The original column definition doesn't have `NOT NULL` flag. The `mergeFieldType` didn't handle the type flag, but the `AggFieldType` did.

**Review by time-and-fate — APPROVED** [review]
LGTM.
But there's more difference between the new and old logic than the bugfix itself. I'm not sure if they are all expected and correct.
I think it's better to have another reviewer more familiar with the types.

**Review by xhebox — APPROVED** [review]

**Pull request merged**